soc: arm: atmel_sam: Rework clock initialization

pdgendt opened the pull request:
Start the clock initialization with resetting first to the MAIN clock, same as the reset value.
This fixes the issue of re-initialising the fast clock, for example after the bootloader.

attie-argentum commented:
Please could you clarify the issue and how this resolves it?

I'm not keen on the "_internal oscillator is disabled, therefore everything is configured correctly and as desired_" assumption here... it may work for you, but it'll very probably bite someone else in subtle ways.

I'd much prefer a "_reset to internal RC oscillator, and reconfigure the external crystal oscillator_" approach, if possible.

pdgendt commented:
> I'd much prefer a "_reset to internal RC oscillator, and reconfigure the external crystal oscillator_" approach, if possible.

I've modified the code to switch to the MAIN clock (reset value) before doing the initialization.

CC: @nandojve 

Two follow-up questions:

- Should the same be done for `sam3x`/`sam4e`/`same70`/`samv71`?
- The `same70` and `samv71` actually have a check, is this preferred over resetting to the MAIN clock:

```c
	/*
	 * Setup main external crystal oscillator if not already done
	 * by a previous program i.e. bootloader
	 */

	if (!(PMC->CKGR_MOR & CKGR_MOR_MOSCSEL_Msk)) {
		// ...
	}
```

bjarki-trackunit commented:
Do it for all of them! Just tested it on ATSAM4E16C with MCUBoot, works perfectly :)

nandojve commented:
Hi @pdgendt ,

If you want to proper fix this you need to take care of more things. All this code was implemented before MCUboot to any SAM device. In fact, the support was added not to long ago but officially we still not support on Zephyr, besides MCUboot already have all necessary details in place to make it available.

In regards of your PR, you need to setup the whole chain of registers to make sure that we can support correctly any bootloader. In the future, LPM could disable the on-chip RC oscillator and then when you select the MCK from it you stuck the device if the system do a call to the reset vector.

That said, you must assume that everything was complete reconfigured and system made a call to the reset vector:

1- Enable the internal OSC with 12MHz
2- Select MOSCSEL to internal OSC
3- Set MCK to MAINCK
4- Set EFC0->EEFC_FMR = EEFC_FMR_FWS(0);
5- Enable Clock Failure Detector
6- Configure all other clocks (except disabling the internal OSC)
7- Set EFC0->EEFC_FMR = EEFC_FMR_FWS(5); /* 120MHz */
8- Set MCK to PPLACK
9- Disable internal OSC to save power

It will be nice if you can update all the series.

pdgendt commented:
Hey @nandojve,

> If you want to proper fix this you need to take care of more things. All this code was implemented before MCUboot to any SAM device. In fact, the support was added not to long ago but officially we still not support on Zephyr, besides MCUboot already have all necessary details in place to make it available.

Understood, I will try to do the changes and test on `sam4s`, for me MCUboot is the needed use case, so official support would be nice. I do need the changes from https://github.com/zephyrproject-rtos/zephyr/pull/64215 for my board.

> It will be nice if you can update all the series.

It would even be nicer to convert the `Kconfig` options to device tree devices and properties, and share common code between the SoCs. But this might be too big of a step to take at once.

pdgendt commented:
> 5- Enable Clock Failure Detector

@nandojve, the datasheet has the following:
<img width="469" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/3170234/9f2388bb-d414-4938-a2eb-2ee60f4e7cd0">

Are we guaranteed the slow RC is active?

Section **29.13 Main Clock Failure Detector** however states it will be activated.
<img width="515" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/3170234/087a92ee-3b97-42c2-9392-7def47509e6c">

Anyway, is there an example on how the failure detection should be done as I'm not entirely sure how this works.

pdgendt committed:
soc: arm: atmel_sam: common: Add PMC API for clock management

Add functions to Atmel SAM SoC PMC API. This is an effort to hide
most of the internal registers used in different SAM families.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
soc: arm: atmel_sam: sam4s: Rework clock_init

Update clock_init for the Atmel SAM4S SoC using the new PMC API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
soc: arm: atmel_sam: sam3x: Rework clock_init

Update clock_init for the Atmel SAM3X SoC using the new PMC API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
soc: arm: atmel_sam: sam4e: Rework clock_init

Update clock_init for the Atmel SAM4E SoC using the new PMC API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
@bjarki-trackunit would you mind verifying the current state for the SAM4E if this still works?

@nandojve I think the only missing part is the clock failure detection now.

pdgendt commented on soc/arm/atmel_sam/common/soc_pmc.h:
```diff
@@ -35,4 +39,228 @@ void soc_pmc_peripheral_disable(uint32_t id);
+ *
+ * @param prescaler the prescaler value.
+ */
+void soc_pmc_mck_set_prescaler(uint32_t prescaler);
```
All these functions can be made `static ALWAYS_INLINE` if desired.

nandojve commented on soc/arm/atmel_sam/sam4s/soc.c:
```diff
@@ -29,126 +30,63 @@
+	 * is the safe setting for all of this SoCs usable frequencies.
 	 */
-	PMC->PMC_FSMR |= PMC_FSMR_LPM;
+	EFC0->EEFC_FMR = EEFC_FMR_FWS(5);
+#if defined(ID_EFC1)
+	EFC1->EEFC_FMR = EEFC_FMR_FWS(5);
```
This should be moved just before line 103. Need review all SoC files.

nandojve commented on soc/arm/atmel_sam/sam4s/soc.c:
```diff
@@ -29,126 +30,63 @@
-	/* Wait for the RC oscillator to be turned off. */
-	while (PMC->PMC_SR & PMC_SR_MOSCRCS) {
-		;
+		if (IS_ENABLED(CONFIG_SOC_ATMEL_SAM4S_WAIT_MODE)) {
+			/*
+			 * Instruct CPU to enter Wait mode instead of Sleep mode to
+			 * keep Processor Clock (HCLK) and thus be able to debug
+			 * CPU using JTAG.
+			 */
+			soc_pmc_enable_waitmode();
+		}
```
_soc_pmc_enable_waitmode();_ is an independent configuration, right? Need review all SoC files.

pdgendt replied:
Every SoC had a similar `Kconfig` for this.

nandojve replied:
Everything that we are doing is extract the duplicated code. If you already want to fix this you can create a generic entry at https://github.com/zephyrproject-rtos/zephyr/blob/915f0fd74c6065b39ab84b2ecafce57000c822a7/soc/arm/atmel_sam/Kconfig#L18

My concern is that the `soc_pmc_enable_waitmode();` may not even be part of clock initialization and could maybe be moved directly to the `z_arm_platform_init`. In your refactor somehow it depends on `CONFIG_SOC_ATMEL_<SOC>_EXT_MAINCK` now which not seems right.

pdgendt replied:
I did not change the logic for this, it already worked that way.

nandojve replied:
Then it is a bug that became obvious now : ) Please, move out clock init to avoid future confusion. It is better if you configure before clock because allow debug attach faster.

nandojve commented on soc/arm/atmel_sam/common/soc_pmc.c:
```diff
@@ -68,3 +70,317 @@ uint32_t soc_pmc_peripheral_is_enabled(uint32_t id)
+}
+#endif /* CONFIG_SOC_SERIES_SAME70 || CONFIG_SOC_SERIES_SAMV71 */
+
+void soc_pmc_mck_set_source(enum soc_pmc_mck_src source)
```
I think makes more sense set the prescaler together at _soc_pmc_mck_set_source(source, prescaler);_

nandojve replied:
I start to believe that this will be the best signature because we don't know what is the current config. So, when switching the master clock it is better already reinforce what it is the prescaler.

pdgendt replied:
@nandojve,

Problem with this, is that the `same70` and `samv71` have both a prescaler and a divider. So the function signature would be different.

How do you prefer that this is handled?

pdgendt replied:
I've merged `set_source` with prescaler and divider, where the divider is only used on supported SoCs.

nandojve commented on soc/arm/atmel_sam/common/soc_pmc.c:
```diff
@@ -68,3 +70,317 @@ uint32_t soc_pmc_peripheral_is_enabled(uint32_t id)
+	}
+
+	/* Change Fast RC oscillator frequency */
+	PMC->CKGR_MOR = (PMC->CKGR_MOR & ~CKGR_MOR_MOSCRCF_Msk)
+			| CKGR_MOR_KEY_PASSWD
+			| (uint32_t)freq;
```
```diff
	PMC->CKGR_MOR = (PMC->CKGR_MOR & ~CKGR_MOR_MOSCRCF_Msk)
-			| CKGR_MOR_KEY_PASSWD
-			| (uint32_t)freq;
+		      | CKGR_MOR_KEY_PASSWD
+		      | (uint32_t)freq;
```
pipe is below equal signal in this case. Review all changes about this.

nandojve reviewed (commented):
Hi @pdgendt ,

Any particular reason to no update same7x samv7x?

pdgendt commented:
> Any particular reason to no update same7x samv7x?

I can try, but the code was already a bit different.

nandojve commented:
> > Any particular reason to no update same7x samv7x?
> 
> I can try, but the code was already a bit different.

If you can incorporate the equal parts is already fine. Then we have same sequence for all series.

pdgendt committed:
soc: arm: atmel_sam: same70: Rework clock_init

Update clock_init for the Atmel SAME70 SoC using the new PMC API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
soc: arm: atmel_sam: samv71: Rework clock_init

Update clock_init for the Atmel SAMV71 SoC using the new PMC API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
@nandojve I think I've made all requested changes, except the `set_source` with prescaler. The datasheet states that the prescaler should be set depending on the source:
<img width="334" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/3170234/0e2f0c61-4b4b-43fc-b07b-87574f0a727d">

For this reason I think it makes sense to keep these separate so we have this freedom in the calling code.

I can only test using SAM4S SoC, please validate the other too.

@bjarki-trackunit Can you test for the SAM4E SoC?

nandojve commented:
Hi @pdgendt ,
Needs rebase : )

nandojve reviewed (dismissed):
Everything working!

pdgendt commented:
@nandojve I don't have a samv71 to test, can you suggest a fix?

nandojve commented:
> @nandojve I don't have a samv71 to test, can you suggest a fix?

Yes, I'll check it soon.

nandojve commented:
Hi @pdgendt ,

Just apply this patch : )

```diff
diff --git a/soc/arm/atmel_sam/common/soc_pmc.h b/soc/arm/atmel_sam/common/soc_pmc.h
index 719d728848..3be251a3d0 100644
--- a/soc/arm/atmel_sam/common/soc_pmc.h
+++ b/soc/arm/atmel_sam/common/soc_pmc.h
@@ -316,6 +316,8 @@ static ALWAYS_INLINE bool soc_pmc_osc_is_ready_main_xtal(void)
  */
 static ALWAYS_INLINE void soc_pmc_switch_mainck_to_xtal(bool bypass, uint32_t xtal_startup_time)
 {
+       soc_pmc_osc_enable_main_xtal(xtal_startup_time);
+
        /* Enable Main Xtal oscillator */
        if (bypass) {
                PMC->CKGR_MOR = (PMC->CKGR_MOR & ~CKGR_MOR_MOSCXTEN)
diff --git a/soc/arm/atmel_sam/samv71/soc.c b/soc/arm/atmel_sam/samv71/soc.c
index 6e82b9fe2f..5b446a8549 100644
--- a/soc/arm/atmel_sam/samv71/soc.c
+++ b/soc/arm/atmel_sam/samv71/soc.c
@@ -45,7 +45,6 @@ static ALWAYS_INLINE void clock_init(void)
                soc_supc_slow_clock_select_crystal_osc();
        }
 
-
        if (IS_ENABLED(CONFIG_SOC_ATMEL_SAMV71_EXT_MAINCK)) {
                /*
                 * Setup main external crystal oscillator.

```

pdgendt commented:
@nandojve applied

bjarki-andreasen reviewed (approved):
Nice work!

nandojve commented:
I made another round with SAM4S, SAM4E and SAMV71. Everything working!
Thank you so much for this work @pdgendt : )
